c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -279,7 +416,9 @@ pub fn visit<'tcx>(
     maps: &mut AtomMaps<'tcx>,
     loans: &mut HashMap<Local, Vec<(Path, Loan, BorrowKind)>>,
     local_ltys: &[LTy<'tcx>],
+    field_permissions: &HashMap<DefId, PermissionSet>,
     mir: &Body<'tcx>,
+    adt_metadata: &AdtMetadataTable<'tcx>,
```

Comment:
`clippy` warns here that there are too many arguments to the function, and there are a lot.  Could some be refactored out similar to `TypeChecker`, which contains most of them already?

Also, what does `visit` do exactly?  What does it visit?  If it visits an entire MIR body, perhaps it can be `visit_body`.

Reply:
The `visit` function is the entry point and the only public definition in this module.  I don't see much difference between passing 9 arguments to a function vs building a struct with 9 fields and then calling a method on it, so I prefer the simpler approach of just calling the function.  The function is also called in only one place IIRC.

The only visitable thing in the argument list is indeed the MIR `Body`.

Reply:
> I don't see much difference between passing 9 arguments to a function vs building a struct with 9 fields and then calling a method on it

The difference is unless an IDE inserts the parameter names (unlike on GitHub, for example), it's much harder to read, as the fields/parameters are unnamed at the call site.  `clippy` has the warning for a reason.  That said, if you don't agree, we should at least add an `#[allow(clippy::too_many_arguments)]`.

> The only visitable thing in the argument list is indeed the MIR `Body`.

Could we rename it `visit_body` then?  That could be done in a separate PR, though, as it's not related to this one.  `visit_body` is a lot clearer to me, though.

Reply:
Presumably the reason this lint is in clippy and not in rustc is that it produces too many false positives - I think this is an example of one.  The function has only a [single call site](https://github.com/immunant/c2rust/blob/f3d2e044fed855a17f7a0601de279f94ae35ecae/c2rust-analyze/src/borrowck/mod.rs#L312) (and isn't likely to gain more), and at that call site, each argument expression uses a local variable with the same name as the parameter.  The struct version would initialize its fields with `tcx: tcx`, `facts: &mut facts`, `local_ltys: &local_ltys`, etc, which isn't really adding information for the reader.

>we should at least add an `#[allow(clippy::too_many_arguments)]`
>
>Could we rename it `visit_body` then?

These both sound good to me.

Reply:
addressed in 2e2b2f3e